src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php

```diff
@@ -75,11 +96,247 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+    {
+        $formBuilder = $this->get('prestashop.core.form.identifiable_object.builder.order_return_form_builder');
+        $formHandler = $this->get('prestashop.core.form.identifiable_object.handler.order_return_form_handler');
+        $gridFactory = $this->get('prestashop.core.grid.factory.order_return_products');
+
+        try {
+            /** @var EditableOrderReturn $editableOrderReturn */
+            $editableOrderReturn = $this->getQueryBus()->handle(
+                new GetOrderReturnForEditing(
+                    $orderReturnId
+                )
+            );
+
+            $form = $formBuilder->getFormFor($orderReturnId);
+            $form->handleRequest($request);
+
+            $result = $formHandler->handleFor($orderReturnId, $form);
+
+            if ($result->getIdentifiableObjectId()) {
+                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
+
+                return $this->redirectToRoute('admin_merchandise_returns_index');
+            }
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+
+            return $this->redirectToRoute('admin_merchandise_returns_index');
+        }
+
+        $allowPrintingOrderReturnPdf = $editableOrderReturn->getOrderReturnStateId() === self::ORDER_RETURN_STATE_WAITING_FOR_PACKAGE_ID;
+
+        $legacyContext = $this->get('prestashop.adapter.legacy.context');
+
+        return $this->render('@PrestaShop/Admin/Sell/CustomerService/OrderReturn/edit.html.twig', [
+            'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
+            'enableSidebar' => true,
+            'layoutTitle' => sprintf($this->trans('Return Merchandise Authorization (RMA)', 'Admin.Orderscustomers.Feature')),
+            'orderReturnForm' => $form->createView(),
+            'editableOrderReturn' => $editableOrderReturn,
+            'orderReturnsProductsGrid' => $this->presentGrid($gridFactory->getGrid($filters)),
+            'allowPrintingOrderReturnPdf' => $allowPrintingOrderReturnPdf,
+            'dateFormat' => $legacyContext->getLanguage()->date_format_lite,
+        ]);
+    }
+
+    /**
+     * @AdminSecurity("is_granted('delete', request.get('_legacy_controller'))", redirectRoute="admin_merchandise_returns_index")
+     *
+     * @param Request $request
+     * @param int $orderReturnId
+     *
+     * @return void
+     *
+     * @throws CoreException
+     */
+    public function generateOrderReturnPdfAction(Request $request, int $orderReturnId): void
+    {
+        $this->get('prestashop.adapter.pdf.order_return_pdf_generator')->generatePDF([$orderReturnId]);
+
+        // When using legacy generator,
+        // we want to be sure that displaying PDF is the last thing this controller will do
+        die();
```

Comment:
Could you return a new BinaryFileResponse rather than a die() ?

Reply:
But that would require to change how OrderReturnPdfGenerator works no? 
So I actually get the file to return with BinaryFileResponse. 

I am not sure if it's a good idea for MerchandiseReturn page to render pdf differently then other pages currently do. Or I am missing something?

Reply:
You're breaking all possible event listeners that are triggered after the action is executed.
You have to return a BinaryFileResponse with proper data in order to view and/or download the item.
I'm really against adding die in new modern controllers :(

Reply:
@matks @PierreRambaud @Progi1984 

As far as I am aware, there is currently no consensus on how to correctly render pdfs in modern controllers. (Even fairly recent OrderController has multiple ways including the one I use here). 

IMO the "correct" way should be established in a separate PR. And then all places where PDF's are rendered this way should be changed (Including merchandise return controller if this PR is merged by that point.)

I agree that changing the way pdf's are rendered and not using die is a good idea, I just don't think this PR is the place to do it.